### PR TITLE
feat: don't report inbound stream upgrade errors to handler

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

When an inbound stream upgrade fails, there isn't a whole lot we can do about that in the handler. In fact, for several errors, we wouldn't even know which specific handler to target, for example, `NegotiationFailed`. Similiarly, in case of an IO error during the upgrade, we don't know which handler the stream was eventually meant to be for.

## Notes & open questions

~This is a breaking change, thus draft for now.~

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/relay/src/behaviour/handler.rs
Comment (thomaseizinger) [inline protocols/relay/src/behaviour/handler.rs:506]: This event can now be removed/deprecated as well.

File: protocols/relay/src/priv_client/handler.rs
Comment (thomaseizinger) [inline protocols/relay/src/priv_client/handler.rs:330]: Same here.

File: protocols/request-response/src/handler.rs
Comment (thomaseizinger) [inline protocols/request-response/src/handler.rs:180]: Same here.

File: protocols/request-response/src/handler.rs
Comment (thomaseizinger) [inline protocols/request-response/src/handler.rs:171]: And here.

File: swarm/src/connection.rs
Comment (thomaseizinger) [inline swarm/src/connection.rs]: While we are at it, remove this `Timer` variant.

File: swarm/src/connection.rs
Comment (mxinden) [inline swarm/src/connection.rs]: Why not log this error on `debug`?

**Review by mxinden — COMMENTED** [review]
:tada: I am **very** much looking forward to this pull request landing! This has been a thorn in my eye for long.

File: swarm/src/connection.rs
Comment (thomaseizinger) [reply swarm/src/connection.rs]: We could I guess!

**Comment by mxinden** [comment]
@thomaseizinger let me know if you want another review here.

**Comment by thomaseizinger** [comment]
> @thomaseizinger let me know if you want another review here.

Go for it. It is only a draft because it is a breaking change.

File: protocols/dcutr/src/handler/relayed.rs
Comment (mxinden) [inline protocols/dcutr/src/handler/relayed.rs:215]: I think we are on the same page when I say that `libp2p-relay` should not be closing the connection based on `pending_error` via `ConnectionHandlerEvent::Close` further below in `poll`. That said, let's not fix it here but along with https://github.com/libp2p/rust-libp2p/issues/3591.

File: protocols/request-response/CHANGELOG.md
Comment (mxinden) [inline protocols/request-response/CHANGELOG.md]:
```suggestion
- Remove `InboundFailure::UnsupportedProtocols` and `InboundFailure::InboundTimeout`.
  This variants is no longer constructed.
```

**Review by mxinden — DISMISSED** [review]
Ready to merge from my end. Great work!

File: protocols/dcutr/src/handler/relayed.rs
Comment (thomaseizinger) [reply protocols/dcutr/src/handler/relayed.rs:215]: Yes, it should be a much easier change once this lands too!

**Review by mxinden — APPROVED** [review]

**Pull request merged**
